REQUIREMENTS.md

```diff
@@ -0,0 +1,160 @@
+# IPLD Requirements
+
+This document outlines parts of IPLD that should and should not be changed to
+ensure the success of future improvements (especially type systems).
+
+**Definitions**
+
+* Block: A block is a chunk of an IPLD DAG, encoded in a format. Blocks have CIDs.
+* Fragment: A piece of an IPLD DAG. Blocks contain fragments.
+* Node: A node is a *point* in an IPLD DAG (object, array, number, etc.).
+* Link: A link is an IPLD Node that points to another IPLD Node.
+* Path: A paths a human readable pointer to an IPLD Node.
+
+## Linked
+
+IPLD must support linking to any IPLD node (even if the node is in the middle of
+a block). That is, IPLD must support arbitrary IPLD paths in links.
+
+**Motivation:** Considering this in the context of programming languages, not
+being able to *store* a pointer to a struct *inside* of another struct would be
+severely limiting.
```

Comment:
This is actually a tricky subject.

Links that land on a Node in the middle of a Block (because they're a tuple of CID+Path, say) are easy _to read_.  They're basically meaningless if we wanted to try to do an _update_ of a value on the far side of such a Link and propagate it.

It'd be nice to have a paragraph extolling the virtues of linking first... and then tangoing with this tricky subject of links to mid-Block Nodes in a later heading.

Reply:
... I should perhaps clarify that when I say "a bit of hurdle", what I mean is "this is such a huge barrier to clear definition that I think it's nonviable until resolved".

Sometimes I'm a master of understatement.

Reply:
In most cases, I'd copy the node out of the block and update it as a new block. If I can use a path to point to something but I can't _store_ that pointer within IPLD, that pointer is pretty useless.

Reply:
So in this model, let's say I start with one block where the nodes inside contain a link to another block and then the link contains several path segments so we reach a mid-block node.  (So far this is all just setup statements.)

If I do an update, using my library that has a neat function which builds a new node for the updated value, and propagates it up copy-on-write style through all the values I pathed over to reach it, and includes replacing any links with updated link hashes...

... then the very first time I use this update feature, I'd end up rewriting the link to be a CID-only link, since I copied(/replaced) that reached node out of the block?

This seems like it would end up with a lot of data that... wouldn't use this feature, because it would tend to shake itself back out of the data anyway.  Links with path segments that reach a mid-block node would only ever be generated by some application which has opinions about that.  Maybe that's not a problem, but it seems interesting.

Reply:
I wonder is this a drawback of trying to abstract away the block entirely? If you retain the concept of block along with its constituent node then wouldn't this be a bit easier to reason about?